include/mfmg/dealii/amge_host.templates.hpp

```diff
@@ -102,25 +81,118 @@ AMGe_host<dim, MeshEvaluator, VectorType>::AMGe_host(
 }
 
 template <int dim, typename MeshEvaluator, typename VectorType>
+template <typename Triangulation>
+std::tuple<std::vector<std::complex<double>>,
+           std::vector<dealii::Vector<double>>,
+           std::vector<typename VectorType::value_type>,
+           std::vector<dealii::types::global_dof_index>>
+AMGe_host<dim, MeshEvaluator, VectorType>::compute_local_eigenvectors(
+    unsigned int n_eigenvectors, double tolerance,
+    Triangulation const &agglomerate_triangulation,
+    std::map<typename dealii::Triangulation<dim>::active_cell_iterator,
+             typename dealii::DoFHandler<dim>::active_cell_iterator> const
+        &patch_to_global_map,
+    MeshEvaluator const &evaluator,
+    typename std::enable_if_t<is_matrix_free<MeshEvaluator>::value &&
+                                  std::is_class<Triangulation>::value,
+                              int>) const
+{
+  dealii::DoFHandler<dim> agglomerate_dof_handler(agglomerate_triangulation);
+
+  using AgglomerateOperator = MatrixFreeAgglomerateOperator<MeshEvaluator>;
+  AgglomerateOperator agglomerate_operator(evaluator, agglomerate_dof_handler);
+
+  auto const diag_elements = agglomerate_operator.get_diag_elements();
+
+  // Compute the eigenvalues and the eigenvectors
+  unsigned int const n_dofs_agglomerate = agglomerate_operator.m();
+  std::vector<std::complex<double>> eigenvalues(n_eigenvectors);
+  // Arpack only works with double not float
+  std::vector<dealii::Vector<double>> eigenvectors(
+      n_eigenvectors, dealii::Vector<double>(n_dofs_agglomerate));
+
+  auto const eigensolver_type =
+      _eigensolver_params.get<std::string>("type", "lancsoz");
+  if (eigensolver_type == "lanczos")
+  {
+    boost::property_tree::ptree lanczos_params;
+    lanczos_params.put("num_eigenpairs", n_eigenvectors);
+    // We are having trouble with Lanczos when tolerance is too tight.
+    // Typically, it results in spurious eigenvalues (so far, only noticed 0).
+    // This seems to be the result of producing too many Lanczos vectors. For
+    // hierarchy_3d tests, any tolerance below 1e-5 (e.g., 1e-6) produces this
+    // problem. Thus, we try to work around it here. It is still unclear how
+    // robust this is.
+    lanczos_params.put("tolerance", std::max(tolerance, 1e-4));
+    lanczos_params.put("max_iterations",
+                       _eigensolver_params.get("max_iterations", 200));
+    lanczos_params.put("percent_overshoot",
+                       _eigensolver_params.get("percent_overshoot", 5));
+    bool is_deflated = _eigensolver_params.get("is_deflated", false);
+    if (is_deflated)
+    {
+      lanczos_params.put("is_deflated", true);
+      lanczos_params.put("num_cycles",
+                         _eigensolver_params.get<int>("num_cycles"));
+      lanczos_params.put(
+          "num_eigenpairs_per_cycle",
+          _eigensolver_params.get<int>("num_eigenpairs_per_cycle"));
+    }
```

Comment:
Create a function that initializes `lanczos_param` so that we don't duplicate code.

Reply:
Done